### ISSUE TITLE
grpo_ref_split.py +178

### ISSUE THREAD

**Issue by qa6300525** [lsdefine/simple_GRPO#3]
per_token_loss = torch.exp(per_token_logps - per_token_logps.detach()) * advantages.unsqueeze(1)
Why use `per_token_logps - per_token_logps.detach()`, what is the significance of this writing?

**Comment by lsdefine** [comment]
I follow the implementation of trl.  Please see this comment.

https://github.com/huggingface/trl/pull/2565#issuecomment-2595837761

**Comment by lijianshe02** [comment]
@lsdefine I am confused with the implementation, the tensor shape of per_token_logps is (B, L), and the tensor shape of advantages.unsqueeze(1) is (B * num_pre_Q, 1),  when they perform the element-wise operation, the result is per_token_logps * advantages[:B, :], is this result be correct?

**Comment by chenlinzhe** [comment]
per_token_logps - per_token_logps.detach() = 0.detach().....

**Comment by Jerry-hyl** [comment]
In the implementation of trl, the policy model only has a single update following each exploration stage. But here, because GRPO only performs a single rollout operation without updating the old policy, the old policy remains the same as the ref policy. However, using detach() will still cause the old policy to change, which is still an error.

**Comment by Jerry-hyl** [comment]
The implementation of trl updates the policy model once it get a new sample, so it changes GRPO into a pure online RL algorithm and doesn't need clip operation to limit its gradient change.
